File: Typescript/AnagramSorter.ts
/* Given a list of words, group anagrams together
Input: str array Output: 2D str array
Constraints: 1 ≤ strs.length ≤ 10⁴, 1 ≤ strs[i].length ≤ 100, & strs[i] consists of lowercase letters
 */

class AnagramSorterTS {
    solution(strs: string[]): string[][] {
        const map: Map<string, string[]> = new Map();

        for (let i = 0; i < strs.length; i++) {
            const word: string = strs[i];
            const count: number[] = new Array(26).fill(0);

            for (let j = 0; j < word.length; j++) {
                count[word.charCodeAt(j) - 97]++;
            }

            let key: string = "";
            for (let j = 0; j < 26; j++) {
                key += "#" + count[j];
            }

            if (!map.has(key)) {
                map.set(key, []);
            }
            map.get(key)!.push(word);
        }

        const result: string[][] = [];
        for (const value of map.values()) {
            result.push(value);
        }

        return result;
    }
}

(function main(){
    const input: string[] = ["eat", "tea", "tan", "ate", "nat", "bat"];
    const sorter = new AnagramSorterTS();
    const grouped: string[][] = sorter.solution(input);

    console.log("Grouped Anagrams:");
    for (const group of grouped) {
        console.log(group);
    }
})();

File: Python/AnagramSorter.py
# Given a list of words, group anagrams together
# Input: str array Output: 2D str array
# Constraints: 1 ≤ strs.length ≤ 10⁴, 1 ≤ strs[i].length ≤ 100, & strs[i] consists of lowercase letters
from typing import List

class AnagramSorter:
    def solution(self, strs: List[str]) -> List[List[str]]:
        groups = {}

        for i in range(len(strs)):
            word = strs[i]
            count = [0] * 26
            
            for j in range(len(word)):
                count[ord(word[j]) - ord('a')] += 1
            
            keyParts = []
            for j in range(26):
                keyParts.append('#')
                keyParts.append(str(count[j]))
            key = ''.join(keyParts)

            if key not in groups:
                groups[key] = []
            groups[key].append(word)

        result = []
        for value in groups.values():
            result.append(value)
        return result

def main():
    input_words = ["eat", "tea", "tan", "ate", "nat", "bat"]

    sorter = AnagramSorter()
    grouped = sorter.solution(input_words)

    print("Grouped Anagrams:")
    for group in grouped:
        print(group)

if __name__ == "__main__":
    main()